### ISSUE TITLE
PrometheusNotIngestingSamples: not firing, when tsbd cannot store metrics

### ISSUE THREAD

**Issue by ulikl** [prometheus/prometheus#7064]
**What happened?**

When the Storage for the TSBD files is not writeable, e.g. due to access problems or device full, the Prometheus continues working, but it cannot store metrics in the TSDB.

The Alert "PrometheusNotIngestingSamples" is not firing.
I can only see the TSDB write problems in the Prometheus logs files and via false alerts for other cluster component, which use absent expressions. E..g
``` 
KubeletDown
expr: absent(up{job="kubelet"}
  == 1)
for: 15m
...
```

This is very confusing.

**Did you expect to see something different?**

I would have expected the Alert "PrometheusNotIngestingSamples" to alert on this case.

This could be detected by converting the alert to an absent expressing. e.g.

```
    - alert: PrometheusNotIngestingSamples
      annotations:
        description: Prometheus {{$labels.namespace}}/{{$labels.pod}} is not ingesting
          samples.
        summary: Prometheus is not ingesting samples.
      expr: |
        absent(prometheus_tsdb_head_samples_appended_total) == 1
      for: 10m
      labels:
        severity: warning
```

**How to reproduce it (as minimally and precisely as possible)**:

The storage path for the TSDB files must be nearly full or not available.

**Environment**

* Prometheus Operator version:

   Occured with v0.34.1
   Any version should be effected.

* Kubernetes version information:

    v1.16.2
   Any version should should be effected.

* Kubernetes cluster kind:

    OpenShift 4.3

* Manifests:
.../manifests/prometheus-rules.yaml:
```
...
    - alert: PrometheusNotIngestingSamples
      annotations:
        description: Prometheus {{$labels.namespace}}/{{$labels.pod}} is not ingesting
          samples.
        summary: Prometheus is not ingesting samples.
      expr: |
        rate(prometheus_tsdb_head_samples_appended_total{job="prometheus-k8s",namespace="monitoring"}[5m]) <= 0
      for: 10m
      labels:
        severity: warning
...
```

**Comment by roidelapluie** [comment]
Thanks for your report. It looks as if this is actually a question about usage and not development. See our FAQ about high availability, too: https://prometheus.io/docs/introduction/faq/#can-prometheus-be-made-highly-available

It makes more sense to ask questions like this on the [prometheus-users mailing list](https://groups.google.com/forum/#!forum/prometheus-users) rather than in a GitHub issue. On the mailing list, more people are available to potentially respond to your question, and the whole community can benefit from the answers provided.

**Comment by ulikl** [comment]
This is for me a bug in standard Alert "PrometheusNotIngestingSamples" provided via this repository file:
 prometheus/documentation/prometheus-mixin/alerts.libsonnet

**Comment by roidelapluie** [comment]
cc @brancz @tomwilkie what is your view on this?

**Comment by brancz** [comment]
I see one of three possibilities:

1) the metric is wrong
2) the alert is wrong
3) another alert is needed

All of which seem like a legitimate issue, the first two being a bug, the last being a feature that I feel is a necessity in the monitoring-mixin.

**Comment by roidelapluie** [comment]
> I see one of three possibilities:
> 
>     1. the metric is wrong
> 
>     2. the alert is wrong
> 
>     3. another alert is needed
> 
> 
> All of which seem like a legitimate issue, the first two being a bug, the last being a feature that I feel is a necessity in the monitoring-mixin (should it be the case).

if TSDB is full how can we still run queries?

**Comment by roidelapluie** [comment]
I am even wondering if alerts still work at all if disk is full

**Comment by brian-brazil** [comment]
I'd presume that it'd a different Prometheus that'd be doing the alerting.

**Comment by brancz** [comment]
> I'd presume that it'd a different Prometheus that'd be doing the alerting.

Agreed.

**Comment by roidelapluie** [comment]
_"I can only see the TSDB write problems in the Prometheus logs files and via false alerts for other cluster component, which use absent expressions."_

That tends to prove that this issue is for a single prometheus. WDYT?

**Comment by ulikl** [comment]
No, its the same prometheus.
When only writing fails, querying still seems to work. Just every instant query returns empty results. At least after 5 minutes are over for the standard lookback delta.

**Comment by ulikl** [comment]
Hi,
 I just tested the latest Prometheus release 2.17.1 regarding disk full behaviour.

If the disk is full at start of Prometheus, than Prometheus doesn't start.
If the disc gets full when its running, the querying , rule evaluation and alert notification continues normally.


Here are the result in detail:

Disk full at 16:45:50Z

extract from prometheus log:
```
ts=2020-03-29T16:45:42.235Z caller=dedupe.go:112 component=remote level=warn remote_name=f133c2 url="http://svc-influxdb:8086/api/v1/prom/write?db=osinfra " msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1585495830 minSendTimestamp=1585500332
ts=2020-03-29T16:45:42.235Z caller=dedupe.go:112 component=remote level=warn remote_name=604dbf url="http://svc-influxdb:8086/api/v1/prom/write?db=osinfra&rp=capa " msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1585495827 minSendTimestamp=1585500332
ts=2020-03-29T16:45:42.647Z caller=dedupe.go:112 component=remote level=error remote_name=d8b3dc url=http://thanos-receiver:10981/api/v1/receive  msg="error tailing WAL" err="unexpected checksum e7cc37c4, expected f076d328"
ts=2020-03-29T16:45:47.656Z caller=dedupe.go:112 component=remote level=info remote_name=d8b3dc url=http://thanos-receiver:10981/api/v1/receive  msg="replaying WAL" queue=d8b3dc
level=warn ts=2020-03-29T16:45:51.858Z caller=manager.go:595 component="rule manager" group=node.rules msg="rule sample appending failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
level=warn ts=2020-03-29T16:45:51.859Z caller=manager.go:595 component="rule manager" group=node.rules msg="rule sample appending failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
level=warn ts=2020-03-29T16:45:51.859Z caller=manager.go:595 component="rule manager" group=node.rules msg="rule sample appending failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
level=warn ts=2020-03-29T16:45:51.860Z caller=manager.go:595 component="rule manager" group=node.rules msg="rule sample appending failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
level=warn ts=2020-03-29T16:45:52.139Z caller=scrape.go:987 component="scrape manager" scrape_pool=kubernetes-kubelet target=https://10.23.15.17:10250/metrics  msg="appending scrape report failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
ts=2020-03-29T16:45:52.234Z caller=dedupe.go:112 component=remote level=warn remote_name=f133c2 url="http://svc-influxdb:8086/api/v1/prom/write?db=osinfra " msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1585495830 minSendTimestamp=1585500342
ts=2020-03-29T16:45:52.235Z caller=dedupe.go:112 component=remote level=warn remote_name=604dbf url="http://svc-influxdb:8086/api/v1/prom/write?db=osinfra&rp=capa " msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1585495827 minSendTimestamp=1585500342
ts=2020-03-29T16:45:52.236Z caller=dedupe.go:112 component=remote level=warn remote_name=d8b3dc url=http://thanos-receiver:10981/api/v1/receive  msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1585495827 minSendTimestamp=1585500342
level=warn ts=2020-03-29T16:45:52.437Z caller=scrape.go:987 component="scrape manager" scrape_pool=node-exporter target=https://10.23.15.27:9100/metrics  msg="appending scrape report failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
level=warn ts=2020-03-29T16:45:57.282Z caller=scrape.go:987 component="scrape manager" scrape_pool=kubernetes-cadvisor target=https://10.23.15.28:10250/metrics/cadvisor  msg="appending scrape report failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
level=warn ts=2020-03-29T16:46:00.038Z caller=scrape.go:987 component="scrape manager" scrape_pool=kubernetes-kubelet target=https://10.23.15.16:10250/metrics  msg="appending scrape report failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
level=warn ts=2020-03-29T16:46:01.118Z caller=manager.go:595 component="rule manager" group=general.rules msg="rule sample appending failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
ts=2020-03-29T16:46:01.653Z caller=dedupe.go:112 component=remote level=error remote_name=604dbf url="http://svc-influxdb:8086/api/v1/prom/write?db=osinfra&rp=capa " msg="error tailing WAL" err="unexpected checksum e7cc37c4, expected f076d328"
ts=2020-03-29T16:46:02.234Z caller=dedupe.go:112 component=remote level=warn remote_name=f133c2 url="http://svc-influxdb:8086/api/v1/prom/write?db=osinfra " msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1585495830 minSendTimestamp=1585500352
ts=2020-03-29T16:46:02.234Z caller=dedupe.go:112 component=remote level=warn remote_name=604dbf url="http://svc-influxdb:8086/api/v1/prom/write?db=osinfra&rp=capa " msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1585495827 minSendTimestamp=1585500352
ts=2020-03-29T16:46:02.234Z caller=dedupe.go:112 component=remote level=warn remote_name=d8b3dc url=http://thanos-receiver:10981/api/v1/receive  msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1585495827 minSendTimestamp=1585500352
level=warn ts=2020-03-29T16:46:05.737Z caller=scrape.go:987 component="scrape manager" scrape_pool=kubernetes-cadvisor target=https://10.23.15.19:10250/metrics/cadvisor  msg="appending scrape report failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
ts=2020-03-29T16:46:06.653Z caller=dedupe.go:112 component=remote level=info remote_name=604dbf url="http://svc-influxdb:8086/api/v1/prom/write?db=osinfra&rp=capa " msg="replaying WAL" queue=604dbf
level=warn ts=2020-03-29T16:46:09.244Z caller=manager.go:595 component="rule manager" group=prometheus.rules msg="rule sample appending failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
level=warn ts=2020-03-29T16:46:10.646Z caller=scrape.go:987 component="scrape manager" scrape_pool=kubernetes-cadvisor target=https://10.23.15.14:10250/metrics/cadvisor  msg="appending scrape report failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
ts=2020-03-29T16:46:12.167Z caller=dedupe.go:112 component=remote level=warn remote_name=f133c2 url="http://svc-influxdb:8086/api/v1/prom/write?db=osinfra " msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1585495830 minSendTimestamp=1585500362
ts=2020-03-29T16:46:12.167Z caller=dedupe.go:112 component=remote level=warn remote_name=604dbf url="http://svc-influxdb:8086/api/v1/prom/write?db=osinfra&rp=capa " msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1585495827 minSendTimestamp=1585500362
ts=2020-03-29T16:46:12.172Z caller=dedupe.go:112 component=remote level=warn remote_name=d8b3dc url=http://thanos-receiver:10981/api/v1/receive  msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1585495827 minSendTimestamp=1585500362
level=warn ts=2020-03-29T16:46:12.816Z caller=scrape.go:987 component="scrape manager" scrape_pool=node-exporter target=https://10.23.15.16:9100/metrics  msg="appending scrape report failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
level=warn ts=2020-03-29T16:46:16.511Z caller=scrape.go:987 component="scrape manager" scrape_pool=kubernetes-kubelet target=https://10.23.15.24:10250/metrics  msg="appending scrape report failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
level=warn ts=2020-03-29T16:46:18.538Z caller=scrape.go:987 component="scrape manager" scrape_pool=node-exporter target=https://10.23.15.28:9100/metrics  msg="appending scrape report failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
level=warn ts=2020-03-29T16:46:19.735Z caller=scrape.go:987 component="scrape manager" scrape_pool=node-exporter target=https://10.23.15.15:9100/metrics  msg="appending scrape report failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
level=warn ts=2020-03-29T16:46:20.494Z caller=scrape.go:987 component="scrape manager" scrape_pool=kubernetes-cadvisor target=https://10.23.15.16:10250/metrics/cadvisor  msg="appending scrape report failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
level=warn ts=2020-03-29T16:46:21.539Z caller=scrape.go:987 component="scrape manager" scrape_pool=node-exporter target=https://10.23.15.23:9100/metrics  msg="appending scrape report failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
ts=2020-03-29T16:46:22.167Z caller=dedupe.go:112 component=remote level=warn remote_name=f133c2 url="http://svc-influxdb:8086/api/v1/prom/write?db=osinfra " msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1585495830 minSendTimestamp=1585500372
ts=2020-03-29T16:46:22.167Z caller=dedupe.go:112 component=remote level=warn remote_name=604dbf url="http://svc-influxdb:8086/api/v1/prom/write?db=osinfra&rp=capa " msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1585495827 minSendTimestamp=1585500372
ts=2020-03-29T16:46:22.173Z caller=dedupe.go:112 component=remote level=warn remote_name=d8b3dc url=http://thanos-receiver:10981/api/v1/receive  msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1585495827 minSendTimestamp=1585500372
level=warn ts=2020-03-29T16:46:23.774Z caller=scrape.go:987 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics  msg="appending scrape report failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
level=warn ts=2020-03-29T16:46:25.637Z caller=manager.go:595 component="rule manager" group=Dashboard_Cluster msg="rule sample appending failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
level=warn ts=2020-03-29T16:46:25.637Z caller=manager.go:595 component="rule manager" group=Dashboard_Cluster msg="rule sample appending failed" err="write to WAL: log samples: write /prometheus/wal/00003944: no space left on device"
```



##Alerts in Prometheus GUI

from http://.../api/v1/alerts 
```json
{"status":"success","data":{"alerts":[{"labels":{"alertname":"DeadMansSwitch","priority":"OK","selfmonitoring":"true","severity":"none"},"annotations":{"description":"This is a DeadMansSwitch meant to ensure that the entire Alerting pipeline is functional.","summary":"Alerting DeadMansSwitch"},"state":"firing","activeAt":"2020-03-29T14:48:01.11767761Z","value":"1e+00"},{"labels":{"alertname":"PrometheusNotIngestingSamples_absent","severity":"warning"},"annotations":{"description":"Prometheus / is not ingesting samples.","summary":"Prometheus is not ingesting samples."},"state":"firing","activeAt":"2020-03-29T16:49:59.306145731Z","value":"1e+00"},{"labels":{"alertname":"KubeAPIDown","severity":"critical"},"annotations":{"message":"KubeAPI has disappeared from Prometheus target discovery."},"state":"firing","activeAt":"2020-03-29T16:51:59.306145731Z","value":"1e+00"}]}}
```


##Outgoing notification calls to alertmanager
still all previously active alerts and these two new

```json
[
  {
    "labels": {
      ...
      "severity": "warning"
    },
    "annotations": {
      "description": "Prometheus / is not ingesting samples.",
      "summary": "Prometheus is not ingesting samples."
    },
    "startsAt": "2020-03-29T16:59:59.306145731Z",
    "endsAt": "2020-03-29T17:17:59.306145731Z",
    "generatorURL": "https://rt-prometheus-prom-monitoring.paasint.consol.de/graph?g0.expr=absent%28prometheus_tsdb_head_samples_appended_total%29+%3D%3D+1&g0.tab=1"
  },
  {
    "labels": {
      ...
      "severity": "critical"
    },
    "annotations": {
      "message": "KubeAPI has disappeared from Prometheus target discovery."
    },
    "startsAt": "2020-03-29T17:07:59.306145731Z",
    "endsAt": "2020-03-29T17:17:59.306145731Z",
    "generatorURL": "https://rt-prometheus-prom-monitoring.paasint.consol.de/graph?g0.expr=absent%28up%7Bjob%3D~%22.%2Aapiserver.%2A%22%7D+%3D%3D+1%29&g0.tab=1"
  }
]
```


but this alert is not firing:

``` 
PrometheusNotIngestingSamples (0 active)

alert: PrometheusNotIngestingSamples
expr: rate(prometheus_tsdb_head_samples_appended_total{job="prometheus-k8s",namespace="openshift-monitoring"}[5m])
  <= 0
for: 10m
labels:
  severity: warning
annotations:
  description: Prometheus {{$labels.namespace}}/{{$labels.pod}} is not ingesting samples.
  summary: Prometheus is not ingesting samples.
```

**Comment by logopk** [comment]
This is not related to the reported Alert bug, but I can confirm that disk full is not discovered and recovered. Results in that timeframe are obviuosly missing. But prometheus continues scraping after the diskspace is freed. In the log is a db.go error "compaction failed"...
_Grafana pulls current results and is happy._
Now once the prometheus service is restarted (even after a day or later), WAL corruption is discovered, the records after corruption are discarded and all data that was displayed before is lost.

So it is absolutely necessary to provide an early alarm to fix the underlying problem, restart prometheus and reduce dataloss to a minimum. (well even better: fix the corruption in the running process ;-) ).

`prometheus      | level=warn ts=2020-03-31T19:53:00.605Z caller=scrape.go:987 component="scrape manager" scrape_pool=nginx target=http://nginxexporter:9113/metrics msg="appending scrape report failed" err="write to WAL: log samples: write /prometheus/wal/00001566: no space left on device"
prometheus      | level=error ts=2020-04-01T05:00:01.811Z caller=db.go:659 component=tsdb msg="compaction failed" err="reload blocks: head truncate failed: create checkpoint: read segments: corruption in segment /prometheus/wal/00001566 at 34996224: unexpected checksum 1dea33dd, expected 6d1ae2de"
...

prometheus      | level=error ts=2020-04-02T07:00:01.802Z caller=db.go:659 component=tsdb msg="compaction failed" err="reload blocks: head truncate failed: create checkpoint: read segments: corruption in segment /prometheus/wal/00001566 at 34996224: unexpected checksum 1dea33dd, expected 6d1ae2de"
prometheus      | level=warn ts=2020-04-02T07:59:27.112Z caller=main.go:512 msg="Received SIGTERM, exiting gracefully..."
prometheus      | level=warn ts=2020-04-02T07:59:30.178Z caller=db.go:608 component=tsdb msg="encountered WAL read error, attempting repair" err="read records: corruption in segment /prometheus/wal/00001566 at 34996224: unexpected checksum 1dea33dd, expected 6d1ae2de"
prometheus      | level=warn ts=2020-04-02T07:59:30.178Z caller=wal.go:353 component=tsdb msg="starting corruption repair" segment=1566 offset=34996224
prometheus      | level=warn ts=2020-04-02T07:59:30.178Z caller=wal.go:361 component=tsdb msg="deleting all segments newer than corrupted segment" segment=1566
prometheus      | level=warn ts=2020-04-02T07:59:30.210Z caller=wal.go:383 component=tsdb msg="rewrite corrupted segment" segment=1566
`

Thanks.

**Comment by brancz** [comment]
I think @brian-brazil and @roidelapluie are right here in that meta monitoring is the solution to this. In both scenarios you should have Prometheus servers that are monitoring each other and a set of Prometheus servers that do nothing but monitor the Prometheus servers. For this scenario you can do extremely precise capacity planning to minimize the risk for running out of disk for example (which btw you should be monitoring separately already as well :slightly_smiling_face:). What Prometheus does here is the best thing it can I believe as it tries to (and seemingly successfully) continue operation at all cost.

**Comment by beorn7** [comment]
As result of our bug scrub, I was asked to vet this.

I'd say the existing alert is correct – as in: Prometheus is still ingesting samples in the described situation where the disk is almost full or otherwise becomes unwriteable. The right way to alert about disks running full, _especially_ if you want it to happen in time before the disk is actually full, is a separate alert based on metrics about the disk state, as you would get it from node_exporter for local disk or from your cloud provider for network storage. (Check out the  [node mixin](https://github.com/prometheus/node_exporter/tree/master/docs/node-mixin) for this kind of alerts.) That's clearly preferred over requiring every instrumented binary that needs disk space to monitor the available disk space on their own.

A different question is if there should be more alerts about various failure modes of the TSDB. We have a bunch of them (like for failed compactions), and once the disk is actually full, I'm sure a few of them will fire. Please follow up if you think there are specific failure modes that should be covered (here or file a new issue). I'll close this issue, though, because I think the `PrometheusNotIngestingSamples` is working as intended, and the early warning about disks running full should be implemented as described above.

**Comment by ulikl** [comment]
Hello @beorn7,
thank your very much for the update.

The problem is that 5 minutes after the disk is full, all queries including the once executed during evaluation of the alert rules return empty.
So in this case no current Prometheus metrics related alerts will fire.

Hence I would like to have an alert like

```
- alert: PrometheusNoCurrentSamples
      annotations:
        description: Prometheus is not returning any samples. Probable cause: disk full or not writable
        summary: Prometheus is not returning any samples
      expr: |
        absent({__name__=~".+"})
      for: 10m
      labels:
        severity: critical
```

So that, in the case that the disk alerts have been ignored, at least I can see one alert related to the current problem.

**Comment by beorn7** [comment]
I wouldn't rely on that behavior as an indicator for full disk. One might very well argue that Prometheus should degrade more gracefully and still evaluate queries properly even if the disk is full.

In different news, I like the idea of running a query that should always return something and then trigger an alert if it doesn't. That alert wouldn't be meant to detect a disk-full condition, but would be more a symptom-based smoke test. If it fires, something must be broken very fundamentally. But I would use a query that is cheaper than asking for all metric names. `absent(up)` sounds like a good candidate.

What does everyone think about that?

**Comment by beorn7** [comment]
Thinking about it, this is not an alert you could run on a separate meta-mon Prometheus. It would need to run on the server itself you are testing. Therefore, it would only work if that server is in a condition where it can still generate and send alerts, but for some reason queries always return an empty result. While that's certainly a condition I would like to be notified about, I'm wondering how frequently it will happen in practice. In any case, this mixin is meant to be for meta-monitoring, to be run on the metamon Prometheus server, _not_ on _every_ server, so even if it cannot harm to have that smoke-test alert on every server, this mixin wouldn't be the place to put it.

**Comment by ulikl** [comment]
That's fine for me.
I would suggest as cheaper query
```
  absent(prometheus_tsdb_head_samples_appended_total)
``` 
as this is more related to the Prometheus functionality itself.
If no targets can be discovery (`absent(up)`)  the Prometheus can still be working fine.

**Comment by ulikl** [comment]
Note: the alerts are still sent, when no metrics can be written.

**Comment by beorn7** [comment]
After even more thinking:

If the disk runs full, all bets are off. While the Prometheus server should be designed with graceful degradation in mind, there aren't that many assumption that you can still make in that situation. However, I'd say any errors in the TSDB due to a full disk should bubble up the error where meaningfully possible and definitely increment an error counter so that ingestion or PromQL evaluation don't appear to succeed while doing nothing (i.e. no ingestion happening, empty results returned). So I'd say we are dealing with a bug in Prometheus rather than something we can fix by improving the monitoring alone.

Could you perhaps reproduce the problem with a current version of Prometheus and double-check if perhaps there are error metrics being incremented or errors being logged while ingestion is failing despite `prometheus_tsdb_head_samples_appended_total` getting incremented or while queries return empty results? With your findings, you could then file a bug against the TSDB code.

**Comment by beorn7** [comment]
> `absent(prometheus_tsdb_head_samples_appended_total)`

This query only makes sense for a Prometheus server that scrapes itself or other Prometheus servers. While many Prometheus servers are set up to scrape themselves, that's not a requirement. In fact, in a larger setup with dedicated meta-monitoring servers and perhaps even a remote storage solution included, it might be preferred to not let each server scrape itself.

**Comment by ulikl** [comment]
I did a test for meta-mon monitoring, i.e. I scraped the metrics from the Prometheus, which got the disc full, in another Prometheus instance
I used the latest Prometheus Container image quay.io/prometheus/prometheus  (v2.19.2) 

In case off disc full the following metrics on the meta-mon Prometheus show failure situations
- prometheus_tsdb_compactions_failed_total: increases, when no new files for compaction can be allocated
- prometheus_tsdb_wal_writes_failed_total: increases, when no new WAL file can be allocated. The Prometheus continues to run normally until the currently allocated wal file is filled.
- rate(prometheus_tsdb_head_samples_appended_total[5m]): 0, when no new WAL file can be allocated

In my test none of the prometheus_tsdb_checkpoint_.*failed_total metrics is non zero. I think the check pointing happens only after the previous failed functionality.

For WAL metrics there are not yet any alerts in https://github.com/prometheus/prometheus/blob/master/documentation/prometheus-mixin/alerts.libsonnet

the following alerts fired:
```
           alert: 'PrometheusTSDBCompactionsFailing',
           expr: |||
              increase(prometheus_tsdb_compactions_failed_total{%(prometheusSelector)s}[3h]) > 0
            ||| % $._config,
...

            alert: 'PrometheusNotIngestingSamples',
            expr: |||
              rate(prometheus_tsdb_head_samples_appended_total{%(prometheusSelector)s}[5m]) <= 0
            ||| % $._config,
...
```

No other current alert fired, but I had no Alertmanager or remote write configured.

**Comment by beorn7** [comment]
This behavior totally makes sense to me. Most importantly, the metamon alerting, as provided in this mixin, works exactly as intended. Remember that metamonitoring is meant to happen from a different Prometheus server (the metamon server). A server cannot really metamonitor itself.

Having said all that, I think I understand now better how Prometheus degrades once the disk is full. Everything is actually working as expected. The expected metrics are incremented (e.g. `prometheus_tsdb_compactions_failed_total`) or not incremented (e.g. `prometheus_tsdb_head_samples_appended_total`) as expected. And no new metrics can be ingested at a certain point. Which explains why evaluations return empty results at some point. That's actually a correct evaluation, because no recent samples have made it to the storage. So by now, I don't see any bug in Prometheus.

Feel free to add an `absent` style alert to your Prometheus. It's fine to detect a particular failure mode in smaller, non-production setups. For "real" metamonitoring, it's not robust enough, and I would not recommend it and thus not add it here to the mixin. Also, it's not really a metamonitoring alert, but a "self monitoring" alert, which this mixin is not about.

**Comment by logopk** [comment]
@beorn7 

would you mind to share a feasible query or alert for a single instance to monitor with Grafana?

For my minor private install I don't see a meta monitoring prom-instance.
Nevertheless I would like to notice the issue (if not on failure but then a little prior to failure).

Thanks.

Peter

**Comment by beorn7** [comment]
As @ulikl suggested, you could set up an alert with the query `absent(prometheus_tsdb_head_samples_appended_total)`.

However, even with a single-node setup, it would be much better to also run a node_exporter on that single node and then use a normal disk-filling-up alert to detect the problem _before_ it hits.

**Comment by logopk** [comment]
@beorn7:

sure I do already. But there have been issues, when the WALs got corrupted and I simply don‘t get any values anymore...

**Comment by beorn7** [comment]
Yes, that's the problem if the only thing that monitors your monitoring is your monitoring itself.

However, that's now really out of scope of this GH issue. For support in your case, I recommend the [prometheus-users mailing list](https://groups.google.com/forum/#!forum/prometheus-users).
